training/bert_model_trainer.py

```diff
@@ -0,0 +1,221 @@
+        self.train_categories_path = os.path.join(self.data_dir, 'train_categories.pkl')
+        self.test_sentences_path = os.path.join(self.data_dir, 'valid_sentences.pkl')
+        self.test_categories_path = os.path.join(self.data_dir, 'valid_categories.pkl')
+        self._load_data()
+        self._load_model()
+
+    def _load_data(self):
+        if os.path.exists(self.train_data_path):
+            self._get_data_from_cache()
+        else:
+            self.get_data()
+
+    @property
+    def train_x(self):
+        return self.train_ids
+
+    @property
+    def train_y(self):
+        return self.train_labels
+
+    @property
+    def valid_x(self):
+        return self.test_ids
+
+    @property
+    def valid_y(self):
+        return self.test_labels
+    
+    # Checks cache for pre-existing model, or instantiates a new one.
+    def _load_model(self):
+        if os.path.exists(self.out_dir):
+            self.model = tf.keras.models.load_model(self.out_dir)
+        else:
+            self.get_model()
+
+    # Instantiates a new model with all the hyperparameters provided
+    # to this class.
+    def get_model(self):
+        self.model = BertKerasModel(self.num_classes, bert_dir=self.bert_dir,
+            max_seq_length=self.max_seq_length, dense_size=self.dense_size,
+            dropout_rate=self.dropout_rate)
+        self.model.build(input_shape=(None, self.max_seq_length))
+        self.model.compile(loss='categorical_crossentropy', optimizer=Adam(lr=self.learning_rate),
+            metrics=['accuracy'])
+
+    # Reads cached data. To avoid re-tokenization, which takes a lot of time.
+    def _get_data_from_cache(self):
+        with open(self.train_data_path, 'rb') as f:
+            self.train_ids = pickle.load(f)
+        with open(self.train_labels_path, 'rb') as f:
+            self.train_labels = pickle.load(f)
+        with open(self.valid_data_path, 'rb') as f:
+            self.test_ids = pickle.load(f)
+        with open(self.valid_labels_path, 'rb') as f:
+            self.test_labels = pickle.load(f)
+        self.num_classes = self.train_labels.shape[1]
+
+    """
+    How to split the data into training and validation sets.
+    This varies between experiments so it may be necessary to
+    override this functionality in subclasses.
+    """
+    def get_train_and_valid_split(self, df):
+        return train_test_split(df, random_state=self.RANDOM_SEED)
+
+    """
+    Function to retrieve training and validation data, as well as parse it.
+    This function is responsible for setting the following attributes:
+
+    self.num_classes, self.train_ids, self.train_labels,
+    self.valid_ids, self.valid_labels
+
+    Any subclass must also define those 5 attributes in get_data if overridden.
+    """
+    def get_data(self):
+        self.tokenizer = create_tokenizer(self.bert_dir)
+
+        df = get_filtered_nyt_data_with_scores(self.DATA_PATH).sample(100)
```

Comment:
is 100 intended here? If so add to a comment, otherwise I think a parameter might work better

Reply:
forgot to remove this, was just using the sample(100) for testing on a small amount of data. updated the PR with the change.